README.md
# flutter_midtrans_cc

Midtrans Credit Card With 3DS handler for Flutter

## Initialize midtrans
```
FlutterMidtransCC midtransCC;

@override
initState(){
    midtransCC = FlutterMidtransCC(
        'midtrans_production_url',
        'midtrans_staging_url', 
        'server_key', 
        'client_key',
        isProduction: false,
    );
    super.initState();
}
```

## Get CC Token
```
midtransCC.getCreditCardToken("Credit card number", "Expiry Date",
                      "Expiry year", "cvv");
```

## Charge payment and redirect to 3DS
```
 midtransCC.midtransCharge("token id", 'custom item id', 'item price', item_details, customer_details);
```

see example for more detailed usage

### Notes: you need to activate Midtrans Web SDK before using this package

CHANGELOG.md
## [1.0.2] - 23/05/2020.

* Edit more detailed readme.md

## [1.0.1] - 23/05/2020.

* Fix pub.dev warnings

## [0.0.1] - 23/05/2020.

* Initial Release


test/flutter_midtrans_cc_test.dart
import 'package:flutter_test/flutter_test.dart';

// import 'package:flutter_midtrans_cc/flutter_midtrans_cc.dart';

void main() {
  test('adds one to input values', () {
    // final calculator = Calculator();
    // expect(calculator.addOne(2), 3);
    // expect(calculator.addOne(-7), -6);
    // expect(calculator.addOne(0), 1);
    // expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/src/Webview3DS.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'package:webview_flutter/webview_flutter.dart';

class Webview3DS extends StatefulWidget {
  final String transactionId;
  final String url;
  final String midtransBaseUrl;
  final String serverKey;
  final Function onClosePressed;
  final Function onCompletedNavigation;

  const Webview3DS(
      {Key key,
      this.transactionId,
      this.url,
      this.onClosePressed,
      this.midtransBaseUrl,
      this.onCompletedNavigation,
      this.serverKey})
      : super(key: key);
  @override
  _Webview3DSState createState() => _Webview3DSState();
}

class _Webview3DSState extends State<Webview3DS> {
  bool isLoading = false;
  String headerText = '';
  Completer<WebViewController> _controller = Completer<WebViewController>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        brightness: Brightness.light,
        backgroundColor: Colors.white,
        title: Text(
          headerText,
          style: TextStyle(color: Colors.black),
        ),
        actions: <Widget>[
          Padding(
            padding: EdgeInsets.only(right: 13),
            child: isLoading == true
                ? CircularProgressIndicator(
                    backgroundColor: Colors.blue,
                  )
                : Container(),
          )
        ],
        leading: GestureDetector(
          onTap: widget.onClosePressed,
          child: Icon(
            Icons.close,
            color: Colors.black,
          ),
        ),
      ),
      body: Container(
        child: WebView(
          javascriptMode: JavascriptMode.unrestricted,
          initialUrl: widget.url,
          onPageStarted: (val) {
            print(val);

            setState(() {
              headerText = val;
              if (headerText.startsWith(
                  'https://api.veritrans.co.id/v2/token/rba/callback')) {
                getTransactStatus().then((response) {
                  var extractedData = json.decode(response.body);

                  if (response.statusCode == 200) {
                    if (extractedData['channel_response_code'] == '0') {
                      widget.onCompletedNavigation();
                    } else {
                      if (extractedData['transaction_status'] == 'deny' &&
                          extractedData['fraud_status'] == "deny") {
                        Navigator.pop(context,
                            "[fraud card] Card denied by system, please try another card");
                      } else {
                        Navigator.pop(
                            context, extractedData['channel_response_message']);
                      }
                    }
                  } else {
                    print(extractedData);
                  }
                });
              }
              isLoading = true;
            });
          },
          onPageFinished: (val) {
            setState(() {
              isLoading = false;
            });
          },
          onWebViewCreated: (WebViewController controller) {
            _controller.complete(controller);

            setState(() {
              isLoading = false;

              controller.getTitle().then((value) {
                setState(() {
                  headerText = value;
                });
              });
              print(controller.currentUrl());
            });
          },
        ),
      ),
    );
  }

  Future<http.Response> getTransactStatus() async {
    String url = widget.midtransBaseUrl + 'v2/${widget.transactionId}/status';

    final response = await http.get(url, headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'Authorization': base64.encode(utf8.encode(widget.serverKey + ':'))
    });

    print(response.statusCode);
    print(response.body);

    return response;
  }
}


lib/flutter_midtrans_cc.dart
library flutter_midtrans_cc;

import 'package:flutter/material.dart';

import 'src/Webview3DS.dart';

import 'dart:convert';
import 'dart:core';
import 'package:http/http.dart' as http;

/// A Calculator.
class FlutterMidtransCC {
  /// enter your production url
  final String midtransProductionUrl;

  /// enter your staging url
  final String midtransStagingUrl;

  /// server_key can be found on your midtrans dashboard
  final String serverKey;

  /// client_key can be found on your midtrans dashboard
  final String clientKey;

  /// if true, production url will be used
  /// if false, staging url will be used
  final bool isProduction;

  FlutterMidtransCC(this.midtransProductionUrl, this.midtransStagingUrl,
      this.serverKey, this.clientKey,
      {this.isProduction = true});

  /// generate credit card token, so we can access 3DS
  ///
  /// this also handle fraud detection and other error
  Future<http.Response> getCreditCardToken(String cardNumber, String expiryDate,
      String expiryYear, String cvv) async {
    String baseUrl =
        isProduction == true ? midtransProductionUrl : midtransStagingUrl;

    String url = baseUrl +
        'v2/token?client_key=$clientKey&card_number=$cardNumber&card_exp_month=$expiryDate&card_exp_year=20$expiryYear&card_cvv=$cvv';

    final response = await http.get(url, headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'Authorization': base64.encode(utf8.encode(serverKey + ':'))
    });

    return response;
  }

  Widget webview3DS(
      {String transactionId,
      String url,
      Function onClosePressed,
      Function onCompleteRedirect}) {
    String baseUrl =
        isProduction == true ? midtransProductionUrl : midtransStagingUrl;
    return Webview3DS(
      midtransBaseUrl: baseUrl,
      serverKey: serverKey,
      onClosePressed: onClosePressed,
      onCompletedNavigation: onCompleteRedirect,
      transactionId: transactionId,
      url: url,
    );
  }

  /// Do charge on your transaction
  ///
  /// how to call midtransCharge:
  ///
  /// List<Map> item_details = [{
  ///   'id': yourCustomPaymentId,
  ///   'price': yourFinalPrice,
  ///   'quantity': yourQuantity,
  ///   'name': yourCustomItemName
  /// },];
  ///
  /// Map customer_details = {
  ///   'first_name': customerFirstName,
  ///   'last_name': customerLastName,
  ///   'email': customerEmail,
  ///   'phone': customerPhone
  /// }
  ///
  /// midtransCharge(yourTokenId, yourOrderId, yourGrossAmount, item_details, customer_details);
  ///
  ///
  Future<http.Response> midtransCharge(String tokenId, String orderId,
      String grossAmount, List<Map> itemDetails, Map customerDetails) async {
    String baseUrl =
        isProduction == true ? midtransProductionUrl : midtransStagingUrl;

    String url = baseUrl + 'v2/charge';
    var encodingBody = json.encode({
      'payment_type': 'credit_card',
      'transaction_details': {'order_id': orderId, 'gross_amount': grossAmount},
      'credit_card': {"token_id": tokenId, "authentication": true},
      'item_details': itemDetails,
      'customer_details': customerDetails
    });

    final response = await http.post(url, body: encodingBody, headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json',
      'Authorization': base64.encode(utf8.encode(serverKey + ':'))
    });

    return response;
  }
}


example/lib/main.dart
import 'dart:convert';

import 'package:flutter/material.dart';
import 'package:flutter_midtrans_cc/flutter_midtrans_cc.dart';
import 'SuccessPage.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Midtrans CC',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: FlutterMidtransCCExample(),
    );
  }
}

class FlutterMidtransCCExample extends StatefulWidget {
  @override
  _FlutterMidtransCCExampleState createState() =>
      _FlutterMidtransCCExampleState();
}

class _FlutterMidtransCCExampleState extends State<FlutterMidtransCCExample> {
  TextEditingController ccNumberController;
  TextEditingController expiryDateController;
  TextEditingController expiryYearController;
  TextEditingController cvvController;
  FlutterMidtransCC midtransCC;

  List<Map> item_details = [
    {'id': 'item1', 'price': 15000, 'quantity': 1, 'name': 'roti bakar'}
  ];

  Map customer_details = {
    'first_name': 'John',
    'last_name': 'Doe',
    'email': 'johndoe@example.com',
    'phone': '08881112233344'
  };

  @override
  void initState() {
    midtransCC = FlutterMidtransCC('midtrans_production_url',
        'midtrans_staging_url', 'server_key', 'client_key',
        isProduction: false);
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      padding: EdgeInsets.symmetric(horizontal: 13, vertical: 13),
      child: ListView(
        children: <Widget>[
          TextFormField(
            controller: ccNumberController,
            autocorrect: false,
            autofocus: false,
            keyboardType: TextInputType.number,
            decoration: InputDecoration(
                hintText: 'credit card number', labelText: 'cc number'),
          ),
          TextFormField(
            controller: expiryDateController,
            autocorrect: false,
            autofocus: false,
            keyboardType: TextInputType.number,
            decoration:
                InputDecoration(hintText: 'expiry date', labelText: 'exp date'),
          ),
          TextFormField(
            controller: expiryYearController,
            autocorrect: false,
            autofocus: false,
            keyboardType: TextInputType.number,
            decoration:
                InputDecoration(hintText: 'expiry year', labelText: 'exp year'),
          ),
          TextFormField(
            controller: cvvController,
            autocorrect: false,
            autofocus: false,
            keyboardType: TextInputType.number,
            decoration: InputDecoration(hintText: 'cvv', labelText: 'cvv'),
          ),
          RaisedButton(
            child: Text(
              'Submit',
              style: TextStyle(color: Colors.white),
            ),
            color: Colors.blue,
            onPressed: () {
              midtransCC
                  .getCreditCardToken(
                      ccNumberController.text,
                      expiryDateController.text,
                      expiryYearController.text,
                      cvvController.text)
                  .then((response) {
                print(response.statusCode);
                print(response.body);

                var extractedData = json.decode(response.body);

                if (response.statusCode == 200) {
                  if (extractedData['status_code'] == '200') {
                    midtransCC
                        .midtransCharge(extractedData['token_id'], '0x224s',
                            '15000', item_details, customer_details)
                        .then((response) {
                      var extractedData = json.decode(response.body);
                      print(response.statusCode);

                      if (response.statusCode == 201 ||
                          response.statusCode == 200) {
                        print(response.body);
                        if (extractedData['status_code'] == "201") {
                          Navigator.push(
                            context,
                            MaterialPageRoute(
                              builder: (context) => midtransCC.webview3DS(
                                  transactionId:
                                      extractedData['transaction_id'],
                                  url: extractedData['redirect_url'],
                                  onClosePressed: () {
                                    Navigator.pop(context);
                                  },
                                  onCompleteRedirect: () {
                                    Navigator.pushReplacement(
                                      context,
                                      MaterialPageRoute(
                                        builder: (context) => SuccessPage(),
                                      ),
                                    );
                                  }),
                            ),
                          ).then((result) {
                            print(result);
                          });
                        }
                      } else {
                        print(response.body);
                      }
                    });
                  } else {
                    for (int i = 0;
                        i < extractedData['validation_messages'].length;
                        i++) {
                      var err = extractedData['validation_messages'].toString();

                      print(err);
                    }
                  }
                }
              });
            },
          )
        ],
      ),
    );
  }
}
